### PR TITLE
[Feature] Support Separated and Occluded COCO metric.

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support Separated and Occluded COCO metric in paper [A Tri-Layer Plugin to Improve Occluded Detection.]( https://arxiv.org/abs/2210.10046)

## Modification

Add OccludedSeparatedCocoDataset

## Usage

Detecting occluded objects still remains a challenge for state-of-the-art object detectors.
We implemented the metric presented in paper [A Tri-Layer Plugin to Improve Occluded Detection](https://arxiv.org/abs/2210.10046) to calculate the recall of separated and occluded masks.

There are two ways to use this metric:

### Offline evaluation

We provide a script to calculate the metric with a dumped prediction file.

First, use the `tools/test.py` script to dump the detection results:

```shell
python tools/test.py ${CONFIG} ${MODEL_PATH} --out results.pkl
```

Then, run the `tools/analysis_tools/coco_occluded_separated_recall.py` script to get the recall of separated and occluded masks:

```shell
python tools/analysis_tools/coco_occluded_separated_recall.py results.pkl --out occluded_separated_recall.json
```

The output should be like this:

```
loading annotations into memory...
Done (t=0.51s)
creating index...
index created!
processing detection results...
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 5000/5000, 109.3 task/s, elapsed: 46s, ETA:     0s
computing occluded mask recall...
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 5550/5550, 780.5 task/s, elapsed: 7s, ETA:     0s
COCO occluded mask recall: 58.79%
COCO occluded mask success num: 3263
computing separated mask recall...
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3522/3522, 778.3 task/s, elapsed: 5s, ETA:     0s
COCO separated mask recall: 31.94%
COCO separated mask success num: 1125
+-----------+--------+-------------+
| mask type | recall | num correct |
+-----------+--------+-------------+
| occluded  | 58.79% | 3263        |
| separated | 31.94% | 1125        |
+-----------+--------+-------------+
Evaluation results have been saved to occluded_separated_recall.json.
```

### Online evaluation

We implemented `OccludedSeparatedCocoDataset` which inherited from the `CocoDataset`.
To evaluate the recall of separated and occluded masks during training, just replace the validation dataset type with `'OccludedSeparatedCocoDataset'` in your config:

```python
data = dict(
    samples_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_train2017.json',
        img_prefix=data_root + 'train2017/',
        pipeline=train_pipeline),
    val=dict(
        type='OccludedSeparatedCocoDataset',  # modify this
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline),
    test=dict(
        type='OccludedSeparatedCocoDataset',  # modify this
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline))
```

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/coco_occluded.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+import mmcv
+import numpy as np
+from mmcv.fileio import load
+from mmcv.utils import print_log
+from pycocotools import mask as coco_mask
+from terminaltables import AsciiTable
+
+from .coco import CocoDataset
+
+
```
Comment (BIGWangYuDong) [inline mmdet/datasets/coco_occluded.py:11]: we should add into DATASET?

File: mmdet/datasets/coco_occluded.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    Args:
+        occluded_ann (str): Path to the occluded coco annotation file.
+        separated_ann (str): Path to the separated coco annotation file.
+    """  # noqa
+
+    def __init__(
+            self,
+            *args,
+            occluded_ann='https://www.robots.ox.ac.uk/~vgg/research/tpod/datasets/occluded_coco.pkl',  # noqa
```
Comment (BIGWangYuDong) [inline mmdet/datasets/coco_occluded.py:37]: Should we give an offline setting?

File: docs/en/useful_tools.md [hunk 3]
```diff
@@ -500,3 +500,79 @@ python tools/analysis_tools/confusion_matrix.py ${CONFIG}  ${DETECTION_RESULTS}
 And you will get a confusion matrix like this:
 
 ![confusion_matrix_example](https://user-images.githubusercontent.com/12907710/140513068-994cdbf4-3a4a-48f0-8fd8-2830d93fd963.png)
+
+## COCO Separated & Occluded Mask Metric
```
Comment (BIGWangYuDong) [inline docs/en/useful_tools.md:504]: Also can update in Chinese version.

File: mmdet/datasets/coco_occluded.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            separated_ann='https://www.robots.ox.ac.uk/~vgg/research/tpod/datasets/separated_coco.pkl',  # noqa
+            **kwargs):
+        super().__init__(*args, **kwargs)
+        self.occluded_ann = load(occluded_ann)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/coco_occluded.py:41]: maybe this setting cannot handle the setting as local path, such as `data_root + 'annotations/xxx.pkl'`?

File: docs/en/useful_tools.md (hunk 3 above)
Comment (RangiLyu) [reply docs/en/useful_tools.md:504]: There is no Chinese doc in 2.x

File: mmdet/datasets/coco_occluded.py (hunk 2 above)
Comment (RangiLyu) [reply mmdet/datasets/coco_occluded.py:50]: These two files are small. Loading from http by default will be much more convenient.

File: docs/en/useful_tools.md [hunk 5]
```diff
@@ -500,3 +500,90 @@ python tools/analysis_tools/confusion_matrix.py ${CONFIG}  ${DETECTION_RESULTS}
...
++-----------+--------+-------------+
+| occluded  | 58.79% | 3263        |
+| separated | 31.94% | 1125        |
++-----------+--------+-------------+
+Evaluation results have been saved to occluded_separated_recall.json.
+```
+
+### Online evaluation
+
+We implemented `OccludedSeparatedCocoDataset` which inherited from the `CocoDataset`.
```
Comment (ZwwWayne) [inline docs/en/useful_tools.md:556]: implemented -> implement

File: mmdet/datasets/coco_occluded.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    """COCO dataset with evaluation on separated and occluded masks which
+    presented in paper `A Tri-Layer Plugin to Improve Occluded Detection.
+
+    <https://arxiv.org/abs/2210.10046>`_.
+
+    Separated COCO and Occluded COCO are automatically generated subsets of
+    COCO val dataset, collecting separated objects and partially occluded
+    objects for a large variety of categories. In this way, we define
+    occlusion into two major categories: separated and partially occluded.
+    - Separation: target object segmentation mask is separated into distinct
```
Comment (hhaAndroid) [inline mmdet/datasets/coco_occluded.py:26]:
```suggestion
    
    - Separation: target object segmentation mask is separated into distinct
```

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**